devdoc/rc_ptr.md

```diff
@@ -0,0 +1,34 @@
+# `rc_ptr` requirements
+
+## Overview
+
+`rc_ptr` is a module that encapsulates reference-counted pointers.
+
+## Exposed API
+```c
+typedef void (*RC_PTR_FREE_FUNC)(void* context);
+
+typedef struct RC_PTR_TAG
+{
+    const void* ptr;
+    RC_PTR_FREE_FUNC free_func;
+} RC_PTR;
+
+THANDLE_TYPE_DECLARE(RC_PTR);
+
+#define PRI_RC_PTR "p"
+
+#define RC_PTR_VALUE(rc) ((rc)->ptr)
+
+#define RC_PTR_AS_VOIDPTR(rc) (((rc) == NULL) ? NULL : (rc)->ptr)
+
+MOCKABLE_FUNCTION(, THANDLE(RC_PTR), rc_ptr_create_with_move_memory, const void*, ptr, RC_PTR_FREE_FUNC, free_func);
+```
+
+### rc_ptr_create_with_move_memory
+```c
+MOCKABLE_FUNCTION(, THANDLE(RC_PTR), rc_ptr_create_with_move_memory, const void*, ptr, RC_PTR_FREE_FUNC, free_func);
```

Comment:
>rc_ptr_create_with_move_memory [](http://example.com/codeflow?start=37&length=30)

Errr... so what does this new module bring that `THANDLE` doesn't have already?

Apparently there's "some" memory allocated by "some" means. Now the module takes that memory and will call "free_func".

But THANDLE does all that already (including allocating the memory).

So the question is why is this new module needed...

Reply:
We need a refcounted `void*` for the `waiter` module. 

Either that or we templatize the entire `waiter` module.

Reply:
Implementing with THANDLE(RC_PTR) for now, will templatize later